Kills /obj/item/melee/transforming, replaces it with a transforming weapon component

MrMelbert opened the pull request:
## About The Pull Request

This PR kills off the `transforming` subtype of `/obj/item/melee` and replaces it with a component to handle the transforming behavior, `/datum/component/transforming`.

The transforming component handles updating the variables of an item when it's transformed. Things like force, sharpness, whetstone force bonus, and attack verbs. Similar to the two-handed component, but instead of transforming into a two-hander it remains a one handed weapon.

The "nemesis" behavior (dealing addition damage to certain factions) of the transforming subtype was moved to the cleaving saw only, since it was the only transforming item that used it. In the future, this can be made into a bespoke element/component as well.

The following weapons and items have been updated to use this component:
- Energy Swords / Sabers / Bananium Energy Sword
- Energy Circular Saw
- Energy Dagger
- Energy Axe
- Toy Energy Sword
- Holographic Energy Sword
- Switchblade
- Advanced Medical Tools (Laser scalpel, Mechanical Pinches, Searing Tool)
- Advanced Engineering Tools (Hand Drill, Jaws of Life / Syndicate Jaws of Life)
- Combat Wrench
- Cleaving Saw
- Telescopic Batons / Contractor Batons
- Roasting Stick
- Telescopic Riot Shield
- Energy Shield / Bananium Energy Shield

This PR also touches up the code around the various above items.

## Why It's Good For The Game

Someone asked me how they'd make their nullrod reskin transform like an energy weapon, and I realized there was no good way besides copy + pasting a ton of code.

Instead of copy + pasting the same `attack_self` sequence of transforming across multiple items, this allows for them all to behave semi-consistently. Which is neat.

## Changelog
:cl: Melbert
refactor: Various toggle/transforming items (energy swords / shields, the cleaving saw, switchblades - to name a few) are now handled via component (the transforming component).
code: Slightly cleaned up the code surrounding various transforming items. Comments, better vars, etc.
fix: The Syndicate Jaws of Life no longer become normal Jaws of Life on use
/:cl:

MrMelbert committed:
WIP: Transforming component

MrMelbert committed:
More progress

MrMelbert committed:
knives and stuff

MrMelbert committed:
few more adjustments

MrMelbert committed:
a few more tools / nulls default

MrMelbert committed:
And a rename

MrMelbert committed:
errors

MrMelbert committed:
comments and tweaks

MrMelbert committed:
verb touch up

MrMelbert committed:
c  baton fix

MrMelbert committed:
Merge branch 'master' into transform_component

MrMelbert committed:
tracking stuff

MrMelbert committed:
comments

MrMelbert committed:
Final adjustments, bananium back to prior type

MrMelbert committed:
c baton fix

tralezab commented on code/datums/components/transforming.dm:
```diff
@@ -0,0 +1,241 @@
+	/// If we get sharpened with a whetstone, save the bonus here for later use if we un/redeploy
+	var/sharpened_bonus
+	/// Callback to be invoked whenever the weapon is transformed.
+	var/datum/callback/on_transform_callback
```
move all the default values above here...

tralezab commented on code/datums/components/transforming.dm:
```diff
@@ -0,0 +1,241 @@
+		clumsy_check = TRUE,
+		list/attack_verb_continuous_on,
+		list/attack_verb_simple_on,
+		datum/callback/on_transform_callback,
```
onto the initial arguments! makes it easier for admins to add "default" transforming items, and even if broken, you don't have to check if you're replacing stuff. continuing with this thought...

tralezab commented on code/datums/components/transforming.dm:
```diff
@@ -0,0 +1,241 @@
+	src.sharpness_on = sharpness_on
+	src.hitsound_on = hitsound_on
+	src.w_class_on = w_class_on
+	src.clumsy_check = clumsy_check
```
you're going to be replacing src values with null without the default values being sent.

tralezab commented on code/datums/components/transforming.dm:
```diff
@@ -0,0 +1,241 @@
+	if(islist(attack_verb_continuous_on))
+		src.attack_verb_continuous_on = attack_verb_continuous_on
+	if(islist(attack_verb_simple_on))
+		src.attack_verb_simple_on = attack_verb_simple_on
```
you're having to check islist without the default values being sent.

tralezab reviewed (commented):
Here's an example of the definitions given to the component being in the initialize (or in case of element, attach)

https://github.com/tgstation/tgstation/blob/e7f846bce2498ef03810870dd12d32b4b638af82/code/datums/components/faction_granter.dm#L18

MrMelbert committed:
defualt arguments

MrMelbert committed:
args

Fikou commented:
holo eswords should use it unless you added them already
also if you update the laser scalpel you should update the rest of the power tools

MrMelbert committed:
holographic energy sword

MrMelbert committed:
tools and further reworking

MrMelbert committed:
Merge branch 'master' into transform_component

MrMelbert committed:
hardsuit again

MrMelbert committed:
icon merge conflict CBT

MrMelbert committed:
More tools, and shields too

MrMelbert committed:
fix saw

MrMelbert committed:
get that + outta here

MrMelbert committed:
no GAGS

MrMelbert committed:
what do you want from me?

MrMelbert committed:
clumsiness checks

MrMelbert committed:
Merge branch 'master' into transform_component

necromanceranne commented:
Could this get added to the edagger as well?

MrMelbert commented:
sure, I by all means have no exhaustive list of things that transform so shout them out as you find them

MrMelbert committed:
- converts the energy dagegr
- pens attack_self now calls parent(bruh)
- renamed some procs, return cancel attack

Twaticus commented:
icons/obj/tools.dmi - the syndie jaws cutter icon seems to be missing 
![image](https://user-images.githubusercontent.com/46540570/128839311-8e77e925-7518-42f6-a7cb-9f36221c2631.png)

MrMelbert commented:
> 
> icons/obj/tools.dmi - the syndie jaws cutter icon seems to be missing

i don't see it, i'll double check but I think it's just a github issue

![image](https://user-images.githubusercontent.com/51863163/128924364-2422882c-1195-4f2a-8387-8f4fa336f781.png)

MrMelbert committed:
Merge branch 'master' into transform_component

Twaticus commented:
damn youre right its there. i fell for the icondiffbot ruse

MrMelbert committed:
Merge branch 'master' into transform_component

MrMelbert committed:
comments / missed qdel

MrMelbert committed:
clown

LemonInTheDark commented on code/datums/components/transforming.dm:
```diff
@@ -0,0 +1,272 @@
+		to_chat(user, span_warning("Wait a bit before trying to use [source] again!"))
+		return
+
+	if(pre_transform_callback)
+		if(!pre_transform_callback.Invoke(user))
+			return
```
You could do this with signals, but this is alright too, just a bit more limiting

MrMelbert replied:
Sure

LemonInTheDark commented on code/datums/components/transforming.dm:
```diff
@@ -0,0 +1,272 @@
+ */
+/datum/component/transforming/proc/do_transform(obj/item/source, mob/user)
+	toggle_active(source)
+	if(on_transform_callback)
+		on_transform_callback.Invoke(user, active)
```
kinda ditto with this

MrMelbert commented on code/modules/mining/lavaland/megafauna_loot.dm:
```diff
@@ -795,101 +795,124 @@
+			var/turf/T = get_step(user_turf, turn(dir_to_target, i))
+			for(var/mob/living/L in T)
+				if(user.Adjacent(L) && L.density)
+					melee_attack_chain(user, L)
```
whoops, accidentally reverted when the files were moved

MrMelbert committed:
accidental reversion

MrMelbert committed:
callbacks -> signals
